src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageActionDescriptorProvider.cs

```diff
@@ -34,45 +35,36 @@ public class PageActionDescriptorProvider : IActionDescriptorProvider
 
         public void OnProvidersExecuting(ActionDescriptorProviderContext context)
         {
-            foreach (var item in _project.EnumerateItems(_pagesOptions.RootDirectory))
+            for (var i = 0; i < _pageMetadataProviders.Count; i++)
             {
-                if (item.FileName.StartsWith("_"))
+                var pageMetadataProvider = _pageMetadataProviders[i];
+                foreach (var item in pageMetadataProvider.EnumeratePageMetadata())
```

Comment:
Yeah this is janky now. Make this work like `ApplicationModelProvider` in core. A list of providers create `PageModel` instances. This class is the orchestrator, and then turns the models into ADs. That's the same way it works in Core, and I think we need that structure here.